core/types/block.go

```diff
@@ -183,6 +184,37 @@ func (r *EpochSnarkData) IsEmpty() bool {
 	return len(r.Signature) == 0
 }
 
+// MarshalJSON marshals as JSON.
+func (r EpochSnarkData) MarshalJSON() ([]byte, error) {
+	type EpochSnarkData struct {
+		Bitmap    hexutil.Bytes
+		Signature hexutil.Bytes
+	}
+	var enc EpochSnarkData
+	enc.Bitmap = r.Bitmap.Bytes()
+	enc.Signature = r.Signature
+	return json.Marshal(&enc)
+}
+
+// UnmarshalJSON unmarshals from JSON.
+func (r *EpochSnarkData) UnmarshalJSON(input []byte) error {
+	type EpochSnarkData struct {
+		Bitmap    hexutil.Bytes
+		Signature hexutil.Bytes
+	}
+	var dec EpochSnarkData
+	if err := json.Unmarshal(input, &dec); err != nil {
+		return err
+	}
+	if dec.Bitmap != nil {
```

Comment:
What should `r.UnmarshalJSON()` do if the received `bitmap` and `signature` are nil, but `r` holds some values for it? Should it:

1) keep the values or
2) overwrite the `r` values with the nil interpretation of it

I have a hunch that it should do 2, but I'm honestly asking. 

Even so, this implementation seems that it does a mix. If bitmap is nil, it won't overwrite the value, but if signature is not, it would, while not returning any error.

Reply:
```
Unmarshaler is the interface implemented by types that can unmarshal a JSON description of themselves. The input can be assumed to be a valid encoding of a JSON value. UnmarshalJSON must copy the JSON data if it wishes to retain the data after returning.

By convention, to approximate the behavior of Unmarshal itself, Unmarshalers implement UnmarshalJSON([]byte("null")) as a no-op.
```

It is not clear what it should do, but I would veer towards something that in most cases would make (marshal(unmarshal(json)) be as idempotent as possible. 

Or maybe I'm just overthinking this and it's not that important

Reply:
You are right, its not that clear.

I think there's only one valid case for calling UnmarshalJson on non empty object, that would be want to re-use a struct for unmarshaling to avoid re-allocating memory.

I think we should disallow code from setting some fields of a struct and expecting the rest to be set by unmarshaling, it seems like a fragile and confusing approach.

So to support that behavior, the unmarshal function would need to set all fields when executed. Interestingly though it looks like that implementation would violate the convention you quoted above, but I wrote a test on go playground to check this and it seems that Unmarshal will overwrite a struct with nil when decoding []byte("null") so it looks like its not a no-op.

https://play.golang.org/p/dHiWmj8X7KD

So based on that I think we can drop the nil checks.

Reply:
Ok so I actually raised an issue on the golang repo because this was not clear to me https://github.com/golang/go/issues/48646